### PR TITLE
[vtadmin] Query vtadmin-api from vtadmin-web with fetch + react-query

### PR DESCRIPTION
**Pull request by doeg**
## Description

**tl;dr:**
This PR implements the humble beginnings of our vtadmin-api HTTP fetchers using native fetch and [react-query](https://react-query.tanstack.com/). The API boundary is one of the spicier parts of any client, particularly when it comes to testing and (strong) typing. To that end, I focused on adding a bunch of tests + commentary -- I would love your feedback on these parts, in particular, if you have any.

Now the usual rambling... 

[react-query](https://react-query.tanstack.com/) is still somewhat new to me, and [Mock Service Workers](https://github.com/mswjs/msw) are brand new to me. I _really_ like how this implementation came together, but I expect a lot of unknown unknowns (for me) here. So I've tried to keep this implementation very modular + flexible in anticipation of us wanting to, say, swap in redux instead of react-query, or use gRPC instead of HTTP. 

I expect the functions in `api/http.ts` to become more general as we add more bindings for other vtadmin-api endpoints. I also have a feeling we'll want to pull out the Mock Service Worker parts of `api/http.test.ts` (especially since we could re-use it as a hermetic/offline development server, which might be interesting). For this PR, though, I tried to keep things straightforward and co-located.

This is also the first use of one (1) environment variable! 

It still doesn't look like much... but! The exciting thing is that vtadmin-web is now issuing HTTP requests against vtadmin-api and populating the UI with ✨ real data ✨ from the [local Vitess + Docker example](https://vitess.io/docs/get-started/local-docker/). 

Screen Shot 2021-01-03 at 2 25 23 PM

A couple of TODOs for me that I'll need to ticket:
- Add CI targets for vtadmin-web unit tests + linting
- Streamline the local development process [described in the README](https://github.com/vitessio/vitess/pull/7239/files#diff-c593dcd79ea9d3ded39a3d33dfdab5cd8959c962adb58ebffc4313cc4affe23fR3)

## Related Issue(s)

- https://github.com/vitessio/vitess/pull/7229

## Checklist
- [ ] Should this PR be backported?
- [ ] Tests were added or are not required
- [ ] Documentation was added or is not required

## Deployment Notes
N/A (yet)

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [ ]  VReplication
- [ ]  Cluster Management
- [ ]  Build

### CODE REVIEW

File: web/vtadmin/src/api/http.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+interface HttpOkResponse {
+    ok: true;
+    result: any;
+}
+
+interface HttpErrorResponse {
+    ok: false;
+}
+
+type HttpResponse = HttpOkResponse | HttpErrorResponse;
```
Comment (ajm188) [inline web/vtadmin/src/api/http.ts:28]: God I love union types so much, go2 pls 🥺

File: web/vtadmin/src/api/http.test.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            const endpoint = `/api/tablets`;
+            mockServerJson(endpoint, { foo: 'bar' });
+
+            expect.assertions(1);
+
+            try {
+                await api.vtfetch(endpoint);
+            } catch (e) {
+                /* eslint-disable jest/no-conditional-expect */
+                expect(e.message).toEqual('invalid http envelope');
```
Comment (ajm188) [inline web/vtadmin/src/api/http.test.ts:104]: One of the potential downsides to doing exact equality checks on error messages means that if we ever add additional context to errors, then we can potentially break a bunch of tests for an unrelated change. Do we want to force that, or does it make sense to `expect().contains()`?

**Review by ajm188 — APPROVED** [review]
I really _really_ like this setup. I have one concern about error testing, but after that this lg2m

File: web/vtadmin/src/api/http.test.ts (hunk 2 above)
Comment (doeg) [reply web/vtadmin/src/api/http.test.ts:104]: Hm, probably the best way is to [extend `Error` with custom subclasses](https://javascript.info/custom-errors). Unfortunately, this is pretty cumbersome to do in JS/TS (especially compared to go 🥺) BUT we can then do assertions like `expect(e.name).toEqual('InvalidHttpEnvelope')` or whatever, which also gives us more flexibility with `e.message`.

I considered this before and didn't do it because I'm uh... well, it's a little bit of work 😬 but it's super worth it, so thanks for pointing this out!

File: web/vtadmin/src/api/http.test.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+// The big downside to mocking or "faking" APIs like vtadmin is that
+// we end up re-implementing some (or all) of vtadmin-api in our test environment.
+// It is, unfortunately, impossible to completely avoid this kind of duplication
+// unless we solely use e2e tests (which have their own trade-offs).
+//
+// That said, our use of protobufjs to validate and strongly type HTTP responses
+// means our fake is more robust than it would be otherwise. Since we are using
+// the exact same protos in our fake as in our real vtadmin-api server, we're guaranteed
+// to have type parity.
+//
```
[the remark this answers is not in the document]
Comment (doeg) [reply web/vtadmin/src/api/http.test.ts:41]: Nah I just like whitespace for multi-paragraph comments. I can remove it though!

[the remark this answers is not in the document]
Comment (ajm188) [reply web/vtadmin/src/api/http.test.ts:41]: No, totally fine!

**Comment by doeg** [comment]
@ajm188 thank you for the great suggestions. 🙏 I made a couple of changes:

- I [defined subclasses of `Error`](https://github.com/vitessio/vitess/pull/7239/files#diff-7ec6b0b56ba07a3d90cf8ec73335b9a8643f748c1a8ea27d0e64cdf2ac42aa4bR31) for anticipated failure scenarios in `api/http.ts` per your comment [here](https://github.com/vitessio/vitess/pull/7239/files/b46e03d41fdb2c6748c2313091fca79bc7e6fe46#r551054270). The tests are [a lot sturdier now](https://github.com/vitessio/vitess/pull/7239/files#diff-06a1fa5d6121c8ea9b4a88c721b543c7d29db4531d0ef3690106a775506904ebR138).

- I [added a test case](https://github.com/vitessio/vitess/pull/7239/files#diff-06a1fa5d6121c8ea9b4a88c721b543c7d29db4531d0ef3690106a775506904ebR128) for when `/api/tablets` returns an `{ ok: false }` response.

- I added some a dumb error placeholder to the UI, which looks silly but shows off how nice react-query's error handling is.

Screen Shot 2021-01-03 at 5 29 43 PM

**Review by ajm188 — APPROVED** [review]

File: web/vtadmin/README.md [hunk 4]
```diff
@@ -1,12 +1,64 @@
...
+				"host": {
+					"hostname": "127.0.0.1:15991"
+				},
+				"tags": ["pool:pool1", "cell:zone1", "extra:tag"]
+			},
+			{
+				"host": {
+					"hostname": "127.0.0.1:15992"
+				},
+				"tags": ["dead-dove-do-not-eat"]
```
[the remark this answers is not in the document]
Comment (doeg) [reply web/vtadmin/README.md:38]: (muffled final_countdown.wav plays in the distance)

File: web/vtadmin/src/api/http.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (doeg) [reply web/vtadmin/src/api/http.test.ts]: Too bad I did it anyway!!! 😈
